### PR TITLE
Expose notify_using_readonly

### PR DESCRIPTION
**Pull request by mjp41**
This exposes a readonly notify using, so that the underlying platform
can map the range of pages readonly into the application.  This improves
performance of external pointer on platforms that support lazy commit
of pages as it can access anything in the range.

### CODE REVIEW

File: src/backend/pagemap.h [hunk 1]
```diff
@@ -159,7 +159,7 @@ namespace snmalloc
 #ifdef SNMALLOC_CHECK_CLIENT
       // Allocate a power of two extra to allow the placement of the
       // pagemap be difficult to guess.
-      size_t additional_size = bits::next_pow2(REQUIRED_SIZE) * 2;
+      size_t additional_size = bits::next_pow2(REQUIRED_SIZE) * 4;
```
Comment (nwf) [inline src/backend/pagemap.h:162]: Unrelated change, but probably fine?

File: src/pal/pal_posix.h [hunk 2]
```diff
@@ -215,6 +215,24 @@ namespace snmalloc
...
+     * Notify platform that we will be using these pages for reading.
+     *
+     * On POSIX platforms, lazy commit means that this is a no-op, unless
+     * we have initially mapped the pages as PROT_NONE.
+     */
+    static void notify_using_readonly(void* p, size_t size) noexcept
+    {
+      SNMALLOC_ASSERT(is_aligned_block<OS::page_size>(p, size));
+
+#ifdef SNMALLOC_CHECK_CLIENT
```
Comment (nwf) [inline src/pal/pal_posix.h:228]: Conditioning on `SNMALLOC_CHECK_CLIENT` inside `notify_using_readonly` seems a little dodgy; at the moment the only call is already under that `ifdef` guard already, but I could imagine cases where we'd be wanting this not under that guard.

File: src/backend/pagemap.h (hunk 1 above)
Comment (mjp41) [reply src/backend/pagemap.h:162]: Not unrelated, if the Pagemap is read-only, then you can try to slide along memory looking for metadata in the pagemap.  This changes it from being a 1/3 chance of hitting a live meta-data entry without a segfault, to 1/5 chance.  If the pages are PROT_NONE, you get one shot and will probably fail, i.e. 1/2^38 or something like that.  So just trying to reduce this as an attack vector.

File: src/pal/pal_posix.h (hunk 2 above)
Comment (mjp41) [reply src/pal/pal_posix.h:228]: Good point, at the moment the Pals behave differently for CHECK_CLIENT and not, with initially mapping memory readonly.  This isn't ideal. I think a larger refactor to make Pals not depend on CHECK_CLIENT makes sense, but then we will need to expose more features from the Pal interface.

Comment (mjp41) [reply src/pal/pal_posix.h:228]: Will refactor based on Teams conversation, and add a second commit.

**Comment by davidchisnall** [comment]
I think this is fine.  I'd also considered adding a permissions template parameter to `notify_using` and `reserve` (in reserve, this could replace `committed`, since no-access is equivalent to not-committed) but I'm happy with this if it's simpler / cleaner.

File: src/pal/pal_posix.h (hunk 2 above)
Comment (mjp41) [reply src/pal/pal_posix.h]: Performed a bit of refactoring to make the semantics, hopefully, clearer.

File: src/pal/pal_apple.h [hunk 3]
```diff
@@ -126,12 +126,11 @@ namespace snmalloc
       while (madvise(p, size, MADV_FREE_REUSABLE) == -1 && errno == EAGAIN)
         ;
 
-#  ifdef SNMALLOC_CHECK_CLIENT
-      // This must occur after `MADV_FREE_REUSABLE`.
-      //
-      // `mach_vm_protect` is observably slower in benchmarks.
-      mprotect(p, size, PROT_NONE);
-#  endif
+      if constexpr (PalEnforceAccess)
```
Comment (davidchisnall) [inline src/pal/pal_apple.h:129]: Braces around body of if statement?

File: src/pal/pal_apple.h [hunk 4]
```diff
@@ -205,7 +205,7 @@ namespace snmalloc
     // Apple's `mmap` doesn't support user-specified alignment and only
     // guarantees mappings are aligned to the system page size, so we use
     // `mach_vm_map` instead.
-    template<bool committed>
+    template<bool state_using>
```
Comment (davidchisnall) [inline src/pal/pal_apple.h:208]: Should this be a tri-state `enum`?  Are there cases where we want to provide read-only access to everything (e.g. the entire pagemap)?

File: src/pal/pal_bsd.h [hunk 5]
```diff
@@ -35,13 +35,13 @@ namespace snmalloc
...
-#if defined(SNMALLOC_CHECK_CLIENT) && !defined(NDEBUG)
+#if !defined(NDEBUG)
       memset(p, 0x5a, size);
 #endif
       madvise(p, size, MADV_FREE);
-#ifdef SNMALLOC_CHECK_CLIENT
-      mprotect(p, size, PROT_NONE);
-#endif
+
+      if constexpr (PalEnforceAccess)
```
Comment (davidchisnall) [inline src/pal/pal_bsd.h:43]: Braces?

File: src/pal/pal_apple.h (hunk 4 above)
Comment (mjp41) [reply src/pal/pal_apple.h:208]: There aren't currently cases.  I propose adding if it is required.

**Pull request merged**